[flutter_tools] Can't build apps targeting Android 13 Tiramisu (API level 33)

<!-- Thank you for using Flutter!

     If you are looking for support, please check out our documentation
     or consider asking a question on Stack Overflow:
      * https://flutter.dev/
      * https://api.flutter.dev/
      * https://stackoverflow.com/questions/tagged/flutter?sort=frequent

     If you have found a bug or if our documentation doesn't have an answer
     to what you're looking for, then fill out the template below. Please read
     our guide to filing a bug first: https://flutter.dev/docs/resources/bug-reports
-->

## Overview

### 

Android 13 Tiramisu (API 33) has recently entered beta and soon (in June) will enter "platform stability" phase ([source](https://developer.android.com/about/versions/13)).

As a developer, I'd like to test whether my application works with API 33 (e.g to test [new notification permission](https://developer.android.com/about/versions/13/changes/notification-permission)).

To do this, I first bumped `targetSdkVersion` and `compileSdkVersion` to `33` in my `<flutter_project_root>/android/app/build.gradle`. It failed with the following error:

```
FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':app:mergeDevDebugAssets'.
> Could not resolve all dependencies for configuration ':app:devDebugRuntimeClasspath'.
   > Could not create task ':app:minifyDevReleaseWithR8'.
      > Failed to find Platform SDK with path: platforms;android-33
```

Then I read [this short tutorial](https://developer.android.com/about/versions/13/setup-sdk#build-config-new-agp) and learned that I have to set `targetSdkVersion` and `compileSdkVersion` to `"Tiramisu"` (I am using AGP 7.0.4 – in `flutter_project_root/android/build.gradle` I have `dependencies {  classpath 'com.android.tools.build:gradle:7.0.4'` /*...*/ }`. After doing so, I got the following error:

```
FAILURE: Build failed with an exception.

* Where:
Script '/Users/bartek/fvm/versions/stable/packages/flutter_tools/gradle/flutter.gradle' line: 461

* What went wrong:
A problem occurred evaluating root project 'android'.
> A problem occurred configuring project ':app'.
   > For input string: ""
```

This caught my attention. I opened the '/Users/bartek/fvm/versions/stable/packages/flutter_tools/gradle/flutter.gradle', went to line 461 and saw [this code](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/gradle/flutter.gradle#L461). 

> For reference, [here's the PR](https://github.com/flutter/flutter/pull/92451) that introduced the `substring(8) as int` on `project.android.compileSdkVersion`

It assumes that `compileSdkVersion` is an `int`, which is not the case for `Tiramisu`.

### Quick fix

I wanted to make the app target API 33 and compile, so I modified the  `/Users/bartek/fvm/versions/stable/packages/flutter_tools/gradle/flutter.gradle`. Actually, I've submitted a PR for this so it's easier to see changes.

After the above modifications, the Gradle build was still failing.

## Steps to Reproduce

<!-- Please include full steps to reproduce so that we can reproduce the problem. -->

1. Create a new flutter project (e.g the default counter app).
2. Set `targetSdkVersion` and `compileSdkVersion` to `"Tiramisu"` in `flutter_project_root/android/app/build.gradle`
3. Execute `flutter run`
4. Watch it fail

**Expected results:** <!-- what did you want to see? -->

I expected my app to build successfully and run on the physical/virtual android device.
**Actual results:** <!-- what did you see? -->

Gradle build failed with the following exception:

```
FAILURE: Build failed with an exception.

* Where:
Script '/Users/bartek/fvm/versions/stable/packages/flutter_tools/gradle/flutter.gradle' line: 461

* What went wrong:
A problem occurred evaluating root project 'android'.
> A problem occurred configuring project ':app'.
   > For input string: ""
```

<details>
<summary>Code sample</summary>

<!--
      Please create a minimal reproducible sample that shows the problem
      and attach it below between the lines with the backticks.

      To create it you can use `flutter create bug` command and update the `main.dart` file.

      Alternatively, you can use https://dartpad.dev/
      which is capable of creating and running small Flutter apps.

      Without this we will unlikely be able to progress on the issue, and because of that
      we regretfully will have to close it.
-->

```dart
```

</details>

<details>
  <summary>Logs</summary>

<!--
      Run your application with `flutter run --verbose` and attach all the
      log output below between the lines with the backticks. If there is an
      exception, please see if the error message includes enough information
      to explain how to solve the issue.
-->

```
```

<!--
     Run `flutter analyze` and attach any output of that command below.
     If there are any analysis errors, try resolving them before filing this issue.
-->

```
```

<!-- Finally, paste the output of running `flutter doctor -v` here. -->

```
$ flutter doctor -v
[✓] Flutter (Channel stable, 3.0.0, on macOS 12.3.1 21E258 darwin-arm, locale en-PL)
    • Flutter version 3.0.0 at /Users/bartek/fvm/versions/stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision ee4e09cce0 (9 days ago), 2022-05-09 16:45:18 -0700
    • Engine revision d1b9a6938a
    • Dart version 2.17.0
    • DevTools version 2.12.2

[✓] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
    • Android SDK at /Users/bartek/androidsdk
    • Platform android-32, build-tools 32.1.0-rc1
    • ANDROID_SDK_ROOT = /Users/bartek/androidsdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.4)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Ultimate Edition (version 2022.1)
    • IntelliJ at /Users/bartek/Applications/JetBrains Toolbox/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 67.0.4
    • Dart plugin version 221.5588

[✓] IntelliJ IDEA Ultimate Edition (version 2022.1)
    • IntelliJ at /Users/bartek/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/221.5080.210/IntelliJ IDEA.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.67.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.40.0

[✓] Connected device (3 available)
    • sdk gphone64 arm64 (mobile) • emulator-5554 • android-arm64  • Android 12 (API 32) (emulator)
    • macOS (desktop)             • macos         • darwin-arm64   • macOS 12.3.1 21E258 darwin-arm
    • Chrome (web)                • chrome        • web-javascript • Google Chrome 101.0.4951.64

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

```
$ sdkmanager --list_installed | grep "platforms;android"
  platforms;android-28                                     | 6            | Android SDK Platform 28                 | platforms/android-28
  platforms;android-29                                     | 5            | Android SDK Platform 29                 | platforms/android-29
  platforms;android-30                                     | 3            | Android SDK Platform 30                 | platforms/android-30
  platforms;android-31                                     | 1            | Android SDK Platform 31                 | platforms/android-31
  platforms;android-32                                     | 1            | Android SDK Platform 32                 | platforms/android-32
  platforms;android-Tiramisu                               | 4            | Android SDK Platform Tiramisu           | platforms/android-Tiramisu
```

```
sdkmanager --list_installed | grep "build-tools"
  build-tools;28.0.3                                       | 28.0.3       | Android SDK Build-Tools 28.0.3          | build-tools/28.0.3
  build-tools;29.0.2                                       | 29.0.2       | Android SDK Build-Tools 29.0.2          | build-tools/29.0.2
  build-tools;30.0.2                                       | 30.0.2       | Android SDK Build-Tools 30.0.2          | build-tools/30.0.2
  build-tools;31.0.0                                       | 31.0.0       | Android SDK Build-Tools 31              | build-tools/31.0.0
  build-tools;32.0.0                                       | 32.0.0       | Android SDK Build-Tools 32              | build-tools/32.0.0
  build-tools;32.0.0-rc1                                   | 32.0.0 rc1   | Android SDK Build-Tools 32-rc1          | build-tools/32.0.0-rc1
  build-tools;32.1.0-rc1                                   | 32.1.0 rc1   | Android SDK Build-Tools 32.1-rc1        | build-tools/32.1.0-rc1
  build-tools;33.0.0-rc4                                   | 33.0.0 rc4   | Android SDK Build-Tools 33-rc4          | build-tools/33.0.0-rc4
```
</details>


<!--
      Consider also attaching screenshots and/or videos to better
      illustrate the issue.

      You can upload them directly on GitHub.
      Beware that video file size is limited to 10MB.
-->

--- a/packages/flutter_tools/gradle/flutter.gradle
+++ b/packages/flutter_tools/gradle/flutter.gradle
@@ -458,7 +458,10 @@ class FlutterPlugin implements Plugin<Project> {
     /** Prints error message and fix for any plugin compileSdkVersion or ndkVersion that are higher than the project. */
     private void detectLowCompileSdkVersionOrNdkVersion() {
         project.afterEvaluate {
-            int projectCompileSdkVersion = project.android.compileSdkVersion.substring(8) as int
+            int projectCompileSdkVersion = Integer.MAX_VALUE // Default to zero if using a preview version as all plugins must be using a lower API version
+            if (project.android.compileSdkVersion.substring(8).isInteger()) {
+                projectCompileSdkVersion = project.android.compileSdkVersion.substring(8) as int
+            }
             int maxPluginCompileSdkVersion = projectCompileSdkVersion
             String ndkVersionIfUnspecified = "21.1.6352462" /* The default for AGP 4.1.0 used in old templates. */
             String projectNdkVersion = project.android.ndkVersion ?: ndkVersionIfUnspecified

--- a/packages/flutter_tools/gradle/flutter.gradle
+++ b/packages/flutter_tools/gradle/flutter.gradle
@@ -461,7 +461,7 @@ class FlutterPlugin implements Plugin<Project> {
             int projectCompileSdkVersion = Integer.MAX_VALUE // Default to int max if using a preview version to skip the sdk check.
             if (project.android.compileSdkVersion.substring(8).isInteger()) { // Stable versions use ints, legacy preview uses string.
                 projectCompileSdkVersion = project.android.compileSdkVersion.substring(8) as int
-            }gi
+            }
             int maxPluginCompileSdkVersion = projectCompileSdkVersion
             String ndkVersionIfUnspecified = "21.1.6352462" /* The default for AGP 4.1.0 used in old templates. */
             String projectNdkVersion = project.android.ndkVersion ?: ndkVersionIfUnspecified

--- a/packages/flutter_tools/gradle/flutter.gradle
+++ b/packages/flutter_tools/gradle/flutter.gradle
@@ -458,10 +458,10 @@ class FlutterPlugin implements Plugin<Project> {
     /** Prints error message and fix for any plugin compileSdkVersion or ndkVersion that are higher than the project. */
     private void detectLowCompileSdkVersionOrNdkVersion() {
         project.afterEvaluate {
-            int projectCompileSdkVersion = Integer.MAX_VALUE // Default to zero if using a preview version as all plugins must be using a lower API version
-            if (project.android.compileSdkVersion.substring(8).isInteger()) {
+            int projectCompileSdkVersion = Integer.MAX_VALUE // Default to int max if using a preview version to skip the sdk check.
+            if (project.android.compileSdkVersion.substring(8).isInteger()) { // Stable versions use ints, legacy preview uses string.
                 projectCompileSdkVersion = project.android.compileSdkVersion.substring(8) as int
-            }
+            }gi
             int maxPluginCompileSdkVersion = projectCompileSdkVersion
             String ndkVersionIfUnspecified = "21.1.6352462" /* The default for AGP 4.1.0 used in old templates. */
             String projectNdkVersion = project.android.ndkVersion ?: ndkVersionIfUnspecified

--- a/packages/flutter_tools/test/integration.shard/flutter_build_preview_sdk.dart
+++ b/packages/flutter_tools/test/integration.shard/flutter_build_preview_sdk.dart
@@ -88,10 +88,7 @@ android {
     }
 
     defaultConfig {
-        // TODO: Specify your own unique Application ID (https://developer.android.com/studio/build/application-id.html).
         applicationId "com.example.buildtest"
-        // You can update the following values to match your application needs.
-        // For more information, see: https://docs.flutter.dev/deployment/android#reviewing-the-build-configuration.
         minSdkVersion flutter.minSdkVersion
         targetSdkVersion flutter.targetSdkVersion
         versionCode flutterVersionCode.toInteger()
@@ -100,8 +97,6 @@ android {
 
     buildTypes {
         release {
-            // TODO: Add your own signing config for the release build.
-            // Signing with the debug keys for now, so `flutter run --release` works.
             signingConfig signingConfigs.debug
         }
     }

--- a/packages/flutter_tools/test/integration.shard/flutter_build_preview_sdk.dart
+++ b/packages/flutter_tools/test/integration.shard/flutter_build_preview_sdk.dart
@@ -40,75 +40,35 @@ void main() {
     'build succeeds targeting string compileSdkVersion',
     () async {
       final File buildGradleFile = exampleAppDir.childDirectory('android').childDirectory('app').childFile('build.gradle');
-      // write a build.gradle with compileSdkVersion as `Tiramisu` which is a string preview version
-      buildGradleFile.writeAsStringSync(r'''
-def localProperties = new Properties()
-def localPropertiesFile = rootProject.file('local.properties')
-if (localPropertiesFile.exists()) {
-    localPropertiesFile.withReader('UTF-8') { reader ->
-        localProperties.load(reader)
-    }
-}
+      // write a build.gradle with compileSdkVersion as `android-Tiramisu` which is a string preview version
+      buildGradleFile.writeAsStringSync(buildGradleFile.readAsStringSync().replaceFirst('compileSdkVersion flutter.compileSdkVersion', 'compileSdkVersion "android-Tiramisu"'), flush: true);
+      expect(buildGradleFile.readAsStringSync(), contains('compileSdkVersion "android-Tiramisu"'));
 
-def flutterRoot = localProperties.getProperty('flutter.sdk')
-if (flutterRoot == null) {
-    throw new GradleException("Flutter SDK not found. Define location with flutter.sdk in the local.properties file.")
-}
+      final ProcessResult result = await processManager.run(<String>[
+        flutterBin,
+        ...getLocalEngineArguments(),
+        'build',
+        'apk',
+        '--debug',
+      ], workingDirectory: exampleAppDir.path);
+      expect(result.stdout, contains('Built build/app/outputs/flutter-apk/app-debug.apk.'));
+      expect(exampleAppDir.childDirectory('build')
+        .childDirectory('app')
+        .childDirectory('outputs')
+        .childDirectory('apk')
+        .childDirectory('debug')
+        .childFile('app-debug.apk').existsSync(), true);
+    },
+  );
 
-def flutterVersionCode = localProperties.getProperty('flutter.versionCode')
-if (flutterVersionCode == null) {
-    flutterVersionCode = '1'
-}
+  test(
+    'build succeeds targeting string compileSdkPreview',
+    () async {
+      final File buildGradleFile = exampleAppDir.childDirectory('android').childDirectory('app').childFile('build.gradle');
+      // write a build.gradle with compileSdkPreview as `Tiramisu` which is a string preview version
+      buildGradleFile.writeAsStringSync(buildGradleFile.readAsStringSync().replaceFirst('compileSdkVersion flutter.compileSdkVersion', 'compileSdkPreview "Tiramisu"'), flush: true);
+      expect(buildGradleFile.readAsStringSync(), contains('compileSdkPreview "Tiramisu"'));
 
-def flutterVersionName = localProperties.getProperty('flutter.versionName')
-if (flutterVersionName == null) {
-    flutterVersionName = '1.0'
-}
-
-apply plugin: 'com.android.application'
-apply plugin: 'kotlin-android'
-apply from: "$flutterRoot/packages/flutter_tools/gradle/flutter.gradle"
-
-android {
-    compileSdkVersion "android-Tiramisu"
-    ndkVersion flutter.ndkVersion
-
-    compileOptions {
-        sourceCompatibility JavaVersion.VERSION_1_8
-        targetCompatibility JavaVersion.VERSION_1_8
-    }
-
-    kotlinOptions {
-        jvmTarget = '1.8'
-    }
-
-    sourceSets {
-        main.java.srcDirs += 'src/main/kotlin'
-    }
-
-    defaultConfig {
-        applicationId "com.example.buildtest"
-        minSdkVersion flutter.minSdkVersion
-        targetSdkVersion flutter.targetSdkVersion
-        versionCode flutterVersionCode.toInteger()
-        versionName flutterVersionName
-    }
-
-    buildTypes {
-        release {
-            signingConfig signingConfigs.debug
-        }
-    }
-}
-
-flutter {
-    source '../..'
-}
-
-dependencies {
-    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
-}
-''', flush: true);
       final ProcessResult result = await processManager.run(<String>[
         flutterBin,
         ...getLocalEngineArguments(),

--- a/packages/flutter_tools/test/integration.shard/flutter_build_preview_sdk.dart
+++ b/packages/flutter_tools/test/integration.shard/flutter_build_preview_sdk.dart
@@ -2,7 +2,6 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
-import 'package:file_testing/file_testing.dart';
 import 'package:flutter_tools/src/base/file_system.dart';
 import 'package:flutter_tools/src/base/io.dart';
 
@@ -117,8 +116,7 @@ dependencies {
         'apk',
         '--debug',
       ], workingDirectory: exampleAppDir.path);
-      print(result.stdout);
-      print(result.stderr);
+      expect(result.stdout, contains('Built build/app/outputs/flutter-apk/app-debug.apk.'));
       expect(exampleAppDir.childDirectory('build')
         .childDirectory('app')
         .childDirectory('outputs')

--- a/packages/flutter_tools/test/integration.shard/flutter_build_preview_sdk.dart
+++ b/packages/flutter_tools/test/integration.shard/flutter_build_preview_sdk.dart
@@ -1,0 +1,135 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:file_testing/file_testing.dart';
+import 'package:flutter_tools/src/base/file_system.dart';
+import 'package:flutter_tools/src/base/io.dart';
+
+import '../src/common.dart';
+import 'test_utils.dart';
+
+void main() {
+  late Directory tempDir;
+  late String flutterBin;
+  late Directory exampleAppDir;
+
+  setUp(() async {
+    tempDir = createResolvedTempDirectorySync('flutter_plugin_test.');
+    flutterBin = fileSystem.path.join(
+      getFlutterRoot(),
+      'bin',
+      'flutter',
+    );
+    exampleAppDir = tempDir.childDirectory('aaa').childDirectory('example');
+
+    processManager.runSync(<String>[
+      flutterBin,
+      ...getLocalEngineArguments(),
+      'create',
+      '--template=plugin',
+      '--platforms=android',
+      'aaa',
+    ], workingDirectory: tempDir.path);
+  });
+
+  tearDown(() async {
+    tryToDelete(tempDir);
+  });
+
+  test(
+    'build succeeds targeting string compileSdkVersion',
+    () async {
+      final File buildGradleFile = exampleAppDir.childDirectory('android').childDirectory('app').childFile('build.gradle');
+      // write a build.gradle with compileSdkVersion as `Tiramisu` which is a string preview version
+      buildGradleFile.writeAsStringSync(r'''
+def localProperties = new Properties()
+def localPropertiesFile = rootProject.file('local.properties')
+if (localPropertiesFile.exists()) {
+    localPropertiesFile.withReader('UTF-8') { reader ->
+        localProperties.load(reader)
+    }
+}
+
+def flutterRoot = localProperties.getProperty('flutter.sdk')
+if (flutterRoot == null) {
+    throw new GradleException("Flutter SDK not found. Define location with flutter.sdk in the local.properties file.")
+}
+
+def flutterVersionCode = localProperties.getProperty('flutter.versionCode')
+if (flutterVersionCode == null) {
+    flutterVersionCode = '1'
+}
+
+def flutterVersionName = localProperties.getProperty('flutter.versionName')
+if (flutterVersionName == null) {
+    flutterVersionName = '1.0'
+}
+
+apply plugin: 'com.android.application'
+apply plugin: 'kotlin-android'
+apply from: "$flutterRoot/packages/flutter_tools/gradle/flutter.gradle"
+
+android {
+    compileSdkVersion "android-Tiramisu"
+    ndkVersion flutter.ndkVersion
+
+    compileOptions {
+        sourceCompatibility JavaVersion.VERSION_1_8
+        targetCompatibility JavaVersion.VERSION_1_8
+    }
+
+    kotlinOptions {
+        jvmTarget = '1.8'
+    }
+
+    sourceSets {
+        main.java.srcDirs += 'src/main/kotlin'
+    }
+
+    defaultConfig {
+        // TODO: Specify your own unique Application ID (https://developer.android.com/studio/build/application-id.html).
+        applicationId "com.example.buildtest"
+        // You can update the following values to match your application needs.
+        // For more information, see: https://docs.flutter.dev/deployment/android#reviewing-the-build-configuration.
+        minSdkVersion flutter.minSdkVersion
+        targetSdkVersion flutter.targetSdkVersion
+        versionCode flutterVersionCode.toInteger()
+        versionName flutterVersionName
+    }
+
+    buildTypes {
+        release {
+            // TODO: Add your own signing config for the release build.
+            // Signing with the debug keys for now, so `flutter run --release` works.
+            signingConfig signingConfigs.debug
+        }
+    }
+}
+
+flutter {
+    source '../..'
+}
+
+dependencies {
+    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
+}
+''', flush: true);
+      final ProcessResult result = await processManager.run(<String>[
+        flutterBin,
+        ...getLocalEngineArguments(),
+        'build',
+        'apk',
+        '--debug',
+      ], workingDirectory: exampleAppDir.path);
+      print(result.stdout);
+      print(result.stderr);
+      expect(exampleAppDir.childDirectory('build')
+        .childDirectory('app')
+        .childDirectory('outputs')
+        .childDirectory('apk')
+        .childDirectory('debug')
+        .childFile('app-debug.apk').existsSync(), true);
+    },
+  );
+}

--- a/packages/flutter_tools/test/integration.shard/flutter_build_preview_sdk.dart
+++ b/packages/flutter_tools/test/integration.shard/flutter_build_preview_sdk.dart
@@ -41,7 +41,10 @@ void main() {
     () async {
       final File buildGradleFile = exampleAppDir.childDirectory('android').childDirectory('app').childFile('build.gradle');
       // write a build.gradle with compileSdkVersion as `android-Tiramisu` which is a string preview version
-      buildGradleFile.writeAsStringSync(buildGradleFile.readAsStringSync().replaceFirst('compileSdkVersion flutter.compileSdkVersion', 'compileSdkVersion "android-Tiramisu"'), flush: true);
+      buildGradleFile.writeAsStringSync(
+        buildGradleFile.readAsStringSync().replaceFirst('compileSdkVersion flutter.compileSdkVersion', 'compileSdkVersion "android-Tiramisu"'),
+        flush: true
+      );
       expect(buildGradleFile.readAsStringSync(), contains('compileSdkVersion "android-Tiramisu"'));
 
       final ProcessResult result = await processManager.run(<String>[
@@ -66,7 +69,10 @@ void main() {
     () async {
       final File buildGradleFile = exampleAppDir.childDirectory('android').childDirectory('app').childFile('build.gradle');
       // write a build.gradle with compileSdkPreview as `Tiramisu` which is a string preview version
-      buildGradleFile.writeAsStringSync(buildGradleFile.readAsStringSync().replaceFirst('compileSdkVersion flutter.compileSdkVersion', 'compileSdkPreview "Tiramisu"'), flush: true);
+      buildGradleFile.writeAsStringSync(
+        buildGradleFile.readAsStringSync().replaceFirst('compileSdkVersion flutter.compileSdkVersion', 'compileSdkPreview "Tiramisu"'),
+        flush: true
+      );
       expect(buildGradleFile.readAsStringSync(), contains('compileSdkPreview "Tiramisu"'));
 
       final ProcessResult result = await processManager.run(<String>[

